rfcs/20200601-tfx-udsl-semantics.md

```diff
@@ -0,0 +1,425 @@
+    asynchronous pipeline. If there are two nodes inside a sub-pipeline that are
+    trying to read the output of a node outside of the sub-pipeline
+    asynchronously, they might get different results. This is demonstrated by
+    `Trainer` reading the output of `EmbeddingGenerator`. Note that this is only
+    available when the parent pipeline is in asynchronous execution mode.
+-   Synchronous inputs. As a comparison, if two nodes in a sub-pipeline are
+    reading from the same synchronous input, it is guaranteed that they will get
+    the same set of artifacts. This is done through snapshotting the inputs at
+    the beginning of the pipeline (as represented by the small barnacle attached
+    to the left side of the sub-pipeline box in the figure above). In our
+    example, `Trainer` and `Transform` are all reading the output of
+    `ExampleGen` as synchronous input. Note that if the parent pipeline is in
+    synchronous execution mode, the nodes inside the sub-pipeline will always
+    read synchronous inputs.
+
+NOTE: By default, a sub-pipeline can be triggered by any newly available
+synchronous input but it will not be triggered by any newly available
+asynchronous input. We will also discuss and provide custom triggering options
+in future designs.
+
+#### Outputs of sub-pipelines
+
+As a symmetry, there will be two flavors for a node outside the sub-pipeline to
+read the outputs of a node inside the sub-pipeline:
+
+-   Asynchronous outputs. This is the same behavior as normal asynchronous data
+    fetching. `TFLiteConverter` above demonstrates this behavior: it can be
+    triggered as soon as `Trainer` in the sub-pipeline produces a new model.
+    Note that this is only available when the parent pipeline is in asynchronous
+    execution mode.
+-   Synchronous outputs. As a comparison, when a node outside of the
+    sub-pipeline tries to read the synchronous outputs of a node inside the
+    sub-pipeline, the outside node will not get the artifacts until all the
+    nodes inside the sub-pipeline finish execution. This is demonstrated by
+    `Pusher` in the example above: It will be able to read a model produced by a
+    `Trainer` execution only when all nodes in the sub-pipeline finish execution
+    for that sub-pipeline run. Similar to synchronous inputs to a sub-pipeline,
+    this is achieved by snapshotting the outputs of nodes inside a sub-pipeline
+    after all nodes finish execution (as represented by the small barnacle
+    attached to the right side of the sub-pipeline box in the figure above).
+
+### Syntax
+
+The example code below demonstrates the syntax for all proposed semantics in
+this RFC. The example code also adopts the same example in the figure above.
+
+```python
+def create_subpipeline(eg, eb):
+  b = tfx.experimental.SubpipelineInputs(
+      inputs={'examples': eg.outputs['examples']},
+      async_inputs={'embedding': eb.outputs['embedding']})
```

Comment:
Wondering if the input/output config could be wrapped into a decorator for the subpipeline function instead? May make it more readable. (maybe decorator for a future proposal as it may also be worth doing a similar thing for pipelines?)

Something like this:


```python
@subpipeline
def create_subpipeline(subpipeline: SubPipeline, examples: SyncInput, embeddings: AsyncInput) -> SubpipelineOutput(async=['model'], sync=['model', 'validation_result']):

  tx = tfx.Transform(
      examples=examples)
  tr = tfx.Trainer(
      examples=examples,
      embedding=embeddings,
      transform_graph=tx.outputs['transform_graph'])
  iv = tfx.InfraValidator(model=tr.outputs['model'])
   
  # This could be avoided if we used automatic component registration 
  subpipeline.add_components([tx, tr, iv])

  return {'model': tr.outputs['model'], 'validation_result': iv.outputs['validation_result']}

eg = tfx.ExampleGen(...)          # Irrelevant parts omitted
eb = tfx.EmbeddingGenerator(...)  # Irrelevant parts omitted
# Function signature modified by decorator?
sp = create_subpipeline(eg.outputs['example'], eb.outputs['embedding'])
p = tfx.Pusher(
    model=sp.outputs['model'],
    validation_result=sp.outputs['validation_result'])
lt = tfx.TFLiteConverter(model=sp.async_outputs['model'])
return pipeline.Pipeline(
    components=[eg, eb, sp, p, lt], execution_mode=ASYNC)

```

Mostly an idea, just leaving here.

Reply:
Thanks @casassg ! This is definitely an improvement on usability. Leaving it open as an AI on us.

/cc @charlesccychen 
/cc @zhitaoli

Reply:
+1 to the suggested usability improvement. Leaving this to @charlesccychen to also acknowedge.